### PR TITLE
new space ruin, the biological research outpost

### PR DESCRIPTION
**Pull request by mc-oofert**
## About The Pull Request

adds this ruin to space ruin pool
this is a shady (as NT always is) bioresearch outpost that got fucked up by an experiment
this has like some puzzle aspect to it since you gotta find keycards and shit and press buttons to unlock shield gates
this ends with you fighting a heart which if you defeat, destroys the blockade that prevents you from entering the outpost vault

also you can no longer literally just cut indestructible grilles or unanchor indestructible windows

### new puzzle elements or something idk
variant of pressure plate that you cannot remove and it sends a puzzle signal
cooler red puzzle doors that look very foreboding or something idk theyre for this ruin
also puzzle blockades, which are indestructible dense objects that are destroyed if they receive a puzzle signal
and also buttons and keycard pads for puzzles

https://github.com/tgstation/tgstation/assets/70376633/c98807ec-1e7b-49c4-a757-cdbb76a1b566

https://github.com/tgstation/tgstation/assets/70376633/9d5d9dd1-5868-44e6-a978-5ea57b30c298

stuff that throws electric shocks in a pattern, ignores insuls and only knocks down, and no you cannot just run past

https://github.com/tgstation/tgstation/assets/70376633/5772917c-a963-48a4-a743-b0f610801d25

### enemies
living floor, it can only attack stuff on top of it and it attacks until the victim is dead
it is invincible to all but a crowbar, and it cannot move, and it remains hidden until a victim is in range

https://github.com/tgstation/tgstation/assets/70376633/aa1d54f6-b259-4e58-9d44-e393d2131acf

living flesh, it can replace your limbs with itself
the conditions for that are; the limb must have 20 or more brute, victim must be alive and dismemberable, the limb may not be torso or head, or the limb may not be living flesh
alternatively it can replace a missing limb
these are all checked with every attack
they have 20 hp
the limbs in question will sometimes act up, while passively draining nutrition, arms will randomly start pulling nearby stuff, legs may step randomly
limbs when detached, turn into mobs and reactivate AI 2 seconds later.
if the host is shocked, all living flesh limbs will detach, or if the host dies they will also do that

https://github.com/tgstation/tgstation/assets/70376633/765cc99e-c800-4efb-aabe-d68817bbd7ae

## Why It's Good For The Game

ruin variety is cool i think
also the other things i added should be useful for other mappers for bitrunning or whatever

also bug bad for that one fix
## Changelog
:cl:
add: living floor, living flesh, and other stuff for the bioresearch outpost ruin
add: bioresearch outpost ruin
fix: you may not defeat indestructible grilles and windows with mere tools
/:cl:

### CODE REVIEW

File: code/modules/mob/living/basic/ruin_defender/flesh.dm
Comment (Jacquerel) [inline code/modules/mob/living/basic/ruin_defender/flesh.dm:134]: Why are we calling this twice?

File: code/modules/mob/living/basic/ruin_defender/living_floor.dm
Comment (Jacquerel) [inline code/modules/mob/living/basic/ruin_defender/living_floor.dm:29]:
```suggestion
	status_flags = GODMODE //nothing but crowbars may kill us
```

File: code/modules/mob/living/basic/ruin_defender/living_floor.dm
Comment (Jacquerel) [inline code/modules/mob/living/basic/ruin_defender/living_floor.dm:90]:
```suggestion
/mob/living/basic/living_floor/attackby(obj/item/weapon, mob/user, params)
	if(weapon.tool_behaviour != TOOL_CROWBAR)
		return ..()
	balloon_alert(user, "prying...")
	playsound(src, 'sound/items/crowbar.ogg', 45, TRUE)
	if(!do_after(user, 5 SECONDS, src))
		return
	new /obj/effect/gibspawner/generic(loc)
	qdel(src)
```
Early returns & that balloon alert is too long.

File: code/modules/surgery/bodyparts/species_parts/misc_bodyparts.dm
Comment (Jacquerel) [inline code/modules/surgery/bodyparts/species_parts/misc_bodyparts.dm:564]: Would be funny to spawn a mob inside of these in case the limb gets added by a bioscrambler or something (and displaced by one attaching itself) but treat this as an optional niche suggestion.

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/basic/ruin_defender/flesh.dm
Comment (mc-oofert) [reply code/modules/mob/living/basic/ruin_defender/flesh.dm:117]: i would have to paste the trait to like every single definition of the meat limb i think

Comment (Jacquerel) [reply code/modules/mob/living/basic/ruin_defender/flesh.dm:117]: Probably true and a gross limitation of our limb system but probably still sensible.

File: code/modules/mob/living/basic/ruin_defender/flesh.dm
Comment (mc-oofert) [reply code/modules/mob/living/basic/ruin_defender/flesh.dm:49]: if im going to be honest i have no clue how prob in Life works to begin with

File: code/game/objects/items/puzzle_pieces.dm
Comment (mc-oofert) [reply code/game/objects/items/puzzle_pieces.dm:330]: i think its worth knowing who swiped if they like opened the gate which contained some horror or something to murder you i think

File: code/modules/mob/living/basic/ruin_defender/flesh.dm
Comment (mc-oofert) [reply code/modules/mob/living/basic/ruin_defender/flesh.dm:138]: oops

File: code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm
Comment (mc-oofert) [reply code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm:58]: im not sure if i understand

File: code/modules/mob/living/basic/ruin_defender/flesh.dm
Comment (mc-oofert) [reply code/modules/mob/living/basic/ruin_defender/flesh.dm:35]: you cant name them

File: code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm
Comment (Jacquerel) [reply code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm:59]: A rule for signals is that they should have only one source, but `COMSIG_GLOB_PUZZLE_COMPLETED` is sent from a whole load of different places.
You've got two options here, which are:
- Investigate a way to make your puzzles respond to a variety of signals as required.
- Make a global proc which fires this signal so that "technically" it is only sent from one place.

Ninja is essentially our headcoder authority in charge of _why_ this is and may or may not tell us that the second solution is shitcode which we also shouldn't do, but it _is_ the simpler option.

Comment (mc-oofert) [reply code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm:59]: so do i just do
```
/proc/puzzle_completed(id)
SEND_GLOBAL_SIGNAL(COMSIG_GLOB_PUZZLE_COMPLETED, id)
```

File: code/modules/mob/living/basic/ruin_defender/flesh.dm
Comment (Jacquerel) [reply code/modules/mob/living/basic/ruin_defender/flesh.dm:35]: I am reasonably sure that you can, you can find a lot of examples of us doing so in the codebase.
```
			AddComponent(/datum/component/traitor_objective_register, bug, \
				fail_signals = list(COMSIG_QDELETING), \
				penalty = telecrystal_penalty)
```

```
AddComponent(/datum/component/scope, range_modifier = 1.7)
```

```
food_item.AddComponent(/datum/component/edible, foodtypes = GROSS)
```

**Comment by carlarctg** [comment]
this is incredible

File: code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm
Comment (Jacquerel) [reply code/modules/mapfluff/ruins/spaceruin_code/meatderelict.dm:59]: Yeah that's basically what I mean

**Review by Wallemations — APPROVED** [review]
looks fine

**Comment by mc-oofert** [comment]
the fixes infact caused a compile error

**Comment by mc-oofert** [comment]
@ninjanomnom should check the subsystem before merging anyhow

**Comment by Loafin34** [comment]
i am somewhat worried about living flesh, the medbay should have the tools to replace limbs so that teams aren't just killed by them.

**Comment by mc-oofert** [comment]
> i am somewhat worried about living flesh, the medbay should have the tools to replace limbs so that teams aren't just killed by them.

they come off if the host is electrocuted or dies

and also literally just go to medbay, explain your situation of being fucked up by parasites and let them do their thing

**Comment by mc-oofert** [comment]
actually defining queue size on literally every single thing sucks so i am going to make it work better soon today

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/queuelinks.dm
Comment (mc-oofert) [reply code/controllers/subsystem/queuelinks.dm:51]: its probably better to only keep this only incrementable incase something inevitably fucks up and makes some ruin not work

**Comment by MMMiracles** [comment]
From the progress I've seen on the discord and my own glances here, the mapping work seems fine. Should be good to merge once code issues are cleared up.

File: code/controllers/subsystem/queuelinks.dm
Comment (ninjanomnom) [inline code/controllers/subsystem/queuelinks.dm:19]: Make this CRASH too

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
Last minor thing on the link system but that part seems good to go

**Pull request merged**
